### PR TITLE
first class functions

### PR DESCRIPTION
**Pull request by Pavel-Durov**
- First-class functions - treating functions as variables.
- Renamed lib struct `Calc` to `YAIWR`

Examples of first-class functions and closures applications:
1. 
```
// Run-time:
//    stdout: 1
//    stderr:

fun f() {
  let x = 0;
  fun g() {
    x = x + 1;
    return x;
  }
  return g;
}

let a = f();
println(a());
```
2.
```
// Run-time:
//    stdout: 1001
//    stderr:

fun f() {
  let x = 1;
  fun g(num) {
    if (num == 0){
      return x + 1;
    } else {
      x = x * num;
      return g;
    }
  }
  
  return g;
}

let a = f();
a(10);
a(10);
a(10);
println(a(0));

```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -6,28 +6,30 @@ use lrpar::{lrpar_mod, LexParseError, NonStreamingLexer};
...
 use ast::AstNode;
 use err::InterpError;
+use hash::HashId;
 
 use crate::instruction::JumpInstruction;
 
-pub struct Calc {
-    fun_store: HashMap<String, Instruction>,
+pub struct YIWR {
+    fun_store: HashMap<u64, Instruction>,
```
Comment (Pavel-Durov) [inline src/lib.rs:26]: I thought about moving function declarations to the Scope as well, but it was complicated, so I left it here for now.

**Comment by ltratt** [comment]
Are the hashes important for correctness?

**Comment by Pavel-Durov** [comment]
Hashing is not essential for correctness. I used it as a mechanism to translate String to numeric representation.
From what I've read, this hashing function should perform well, but I could've used incremental numeric values instead or leave Strings. YAIWR correctness shouldn't be affected either way.

That's what I understood by "compile the variable names to integers" from:
https://chat.soft-dev.org/softdev/pl/maqe6cwi83bk8gxm6zwk5cprdo
Added it to my TODO list in https://github.com/softdevteam/pavel.yaiwr/pull/13 PR.

**Comment by ltratt** [comment]
> I used it as a mechanism to translate String to numeric representation.

Well, yes :) What do you use the numeric representation for?

**Comment by Pavel-Durov** [comment]
I use it for variables and function id store.
If I have an AST node: `Declare { id: "pavel", value: 1 }`
It will be stored it in scope hashmap as: `{ "5953638110699698698":  1 }` instead of `{ "pavel":  1 }` where `5953638110699698698` is a hash of "pavel" string.

**Comment by Pavel-Durov** [comment]
Is it a bad idea?

**Comment by vext01** [comment]
Not necessarily, although I didn't expect you to have to teach Rust how to hash a string from scratch. What's up with `HashMap<String, ...>`?

Another approach is to convert all variables into a unique integer ID in your bytecode. You would have to do extra work to get the variable name back if you need it (e.g.) to print a human-readable error.

**Comment by vext01** [comment]
I guess I'm also surprised that you had to change the structure of your variable store to implement first-class funcs.

**Comment by Pavel-Durov** [comment]
Well, I didn't have to. I just did. 
Nothing fundamental was changed - instead of using String as keys in variables and function hashmaps, now it's u64. 

I can remove the hashing bit. It's not really important.

What's confusing? That I changed the stores to have `u64` keys or the fact that I used the hash function to convert Strings to `u64`?

**Comment by Pavel-Durov** [comment]
bors try

File: src/instruction.rs [hunk 2]
```diff
@@ -116,6 +121,55 @@ pub enum Instruction {
         alternative: Option<Vec<Instruction>>,
     },
 }
+impl PartialEq for Instruction {
```
Comment (vext01) [inline src/instruction.rs:124]: This couldn't be derived?

Comment (Pavel-Durov) [reply src/instruction.rs:124]: You are right! Im sure it was complaining about `Rc` or something else not implementing `PartialEq`. But now when I tried it it's fine :exploding_head: 
I'll rewrite :)

Comment (Pavel-Durov) [reply src/instruction.rs]: removed :point_right: 21b930426c79df8edf415613756b64514b04da4e

**Comment by vext01** [comment]
Please squash

**Comment by Pavel-Durov** [comment]
squashed :point_right: eab98cedd21f5c588e20bea24953ccc7ecc3020c

**Pull request merged**